### ISSUE TITLE
Find API picks the wrong workspace size from the DB. 

### ISSUE THREAD

**Issue by amberhassaan** [ROCmSoftwarePlatform/MIOpen#2500]
Observed:
`ConvDescriptor::GetWorkspaceSize()` searches the DB first and returns a workspace size of 0. then FindConvolution finds an algorithm and picks a CK based implicit GEMM solver that is not in the DB and needs workspace but we pass workspace size of 0 that we obtained from the DB.

Details: 
To reproduce run (or debug) `test_conv3d --float --all --limit 2` from PR #2429 commit id: f7c5feaa62d4ad30e87199fe5c7feb2f359a6dd4
I see the following in `test/conv_common.hpp`: 
line 1572: We call `filter.GetWorkspaceSize()` and obtain 0 from the database
line 1580: We find an algorithm for convolution using a workspace size of 0
line 1602: We choose the `ConvHipImplicitGemm3DGroupBwdXdlops` as the solver but it needs workspace, hence we hit an assert on the workspace pointer (need debug build for this). 

CC: @JehandadKhan , @junliume . 

PS: I discussed this in detail with Jehandad in a meeting and he confirmed that this is a bug due to the Find API short-circuiting the search from the DB (src/convolution.cpp lines 412-434 instead of enumerating solvers' GetWorkspaceSize (src/convolution.cpp lines 439-460)

**Comment by atamazov** [comment]
@amberhassaan @JehandadKhan @DrizztDoUrden  AFAICS the issue requires some more analysis.

> Find API short-circuiting the search from the DB (src/convolution.cpp lines 412-434 instead of enumerating solvers' GetWorkspaceSize (src/convolution.cpp lines 439-460)

This "short-circuiting" is an implementation of Hybrid, Dynamic Hybrid and Fast Find modes.

If GWSS finds a record in the db and returns workspace size from it, AND that happens somewhere in src/convolution.cpp lines 412-434, then the subsequent Find call should return the solver from the same record. **_If database is not corrupt or outdated, then these values (required workspace size and returned solver name) should be in sync._**

> line 1602: We choose the ConvHipImplicitGemm3DGroupBwdXdlops as the solver...

How do we "choose" this solver? Is it so that we simply select the solver returned by by the Find call?

Note that at line 1602 we not only get the solver name, but also the required workspace size, and we use the latter value for allocating workspace, see below. You may wish to fix the description of the issue as it give false impression that it is GWSS that returns incorrect value is the reason of the issue but this is not so (see comment marked with `\section ffind_gwss_why_not_0` in convolution.cpp).

> ...but it needs workspace, hence we hit an assert on the workspace pointer (need debug build for this).

The Find call returns required workspace size in `perf.memory` and that value is used to resize the workspace at line 1598 of conv_common.hpp. It seems that Find returns wrong workspace size and that needs to be fixed.

@amberhassaan I would appreciate if you provide a log taken with
```bash
export MIOPEN_ENABLE_LOGGING=1
export MIOPEN_ENABLE_LOGGING_CMD=1
export MIOPEN_LOG_LEVEL=6
```

**Comment by atamazov** [comment]
@amberhassaan Please also provide the faulty problem config.

**Comment by atamazov** [comment]
@amberhassaan @JehandadKhan @DrizztDoUrden I think the reason of the issue is a bug in PR #2429. The updated `ConvHipImplicitGemm3DGroupBwdXdlops` solver requires workspace but it is still reusing the default GetWorkspaceSize from the base class which returns 0.

**Comment by atamazov** [comment]
@amberhassaan @JehandadKhan @junliume I highly recommend avoiding creating tickets that require some in-progress PR to reproduce. If an issue cannot be reproduced with clean develop, then normally the issue should not be opened and assigned. Of course, exceptions are possible, e.g. if the issue is carefully investigated and its root reason is absolutely clear (so the PR is only a helper that allows for to reproduce the bug easier).

**Comment by amberhassaan** [comment]
> @amberhassaan @JehandadKhan @DrizztDoUrden I think the reason of the issue is a bug in PR #2429. The updated `ConvHipImplicitGemm3DGroupBwdXdlops` solver requires workspace but it is still reusing the default GetWorkspaceSize from the base class which returns 0.

@atamazov : my bad, I had some local commits that I had not pushed out yet, so PR #2429 and the commit id were wrong in the issue description. I have since corrected that. 

You'll see that the solvers have been updated to provide a `GetWorkspaceSize()` with the right size in bytes. 

This is an issue that prevents this PR from passing the CI hence I opened the issue on a WIP PR. I will upload log files by early next week if not end of this week.
